sdk/src/wallet/migration/chrysalis.rs

```diff
@@ -0,0 +1,314 @@
+// Copyright 2023 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{
+    collections::{HashMap, HashSet},
+    convert::TryInto,
+    io::Read,
+    path::{Path, PathBuf},
+    str::FromStr,
+};
+
+use crypto::{
+    ciphers::{chacha::XChaCha20Poly1305, traits::Aead},
+    macs::hmac::HMAC_SHA512,
+};
+use rocksdb::{IteratorMode, DB};
+use serde::Serialize;
+use serde_json::Value;
+use zeroize::Zeroizing;
+
+use crate::{
+    client::{constants::IOTA_COIN_TYPE, storage::StorageAdapter, Password},
+    types::block::address::Bech32Address,
+    wallet::{
+        migration::{MigrationData, MIGRATION_VERSION_KEY},
+        storage::{
+            constants::{
+                ACCOUNTS_INDEXATION_KEY, ACCOUNT_INDEXATION_KEY, CHRYSALIS_STORAGE_KEY, SECRET_MANAGER_KEY,
+                WALLET_INDEXATION_KEY,
+            },
+            StorageManager,
+        },
+        Error, Result,
+    },
+};
+
+#[derive(Debug, Clone, Serialize)]
+#[serde(rename_all = "camelCase")]
+struct AccountAddress {
+    address: Bech32Address,
+    key_index: u32,
+    internal: bool,
+    used: bool,
+}
+
+#[derive(Clone, Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub(crate) struct AccountDetailsDto {
+    pub(crate) index: u32,
+    coin_type: u32,
+    alias: String,
+    public_addresses: Vec<AccountAddress>,
+    internal_addresses: Vec<AccountAddress>,
+    addresses_with_unspent_outputs: Vec<Value>,
+    outputs: HashMap<String, Value>,
+    locked_outputs: HashSet<String>,
+    unspent_outputs: HashMap<String, Value>,
+    transactions: HashMap<String, Value>,
+    pending_transactions: HashSet<String>,
+    incoming_transactions: HashMap<String, Value>,
+    native_token_foundries: HashMap<String, Value>,
+}
+
+pub async fn migrate_db_chrysalis_to_stardust(
+    storage_path: impl Into<PathBuf> + Send,
+    password: Option<Password>,
+    new_db_encryption_key: impl Into<Option<Zeroizing<[u8; 32]>>> + Send,
```

Comment:
Shouldn't this be a password?

Reply:
No, https://github.com/iotaledger/iota-sdk/blob/5cfc4aa4cb8ef14708cd42ab313198536f6ee4fa/sdk/src/wallet/storage/options.rs#L44

Reply:
Well yeah sure, but `migrate_db_chrysalis_to_stardust` is a bit more user-facing, we could convert a password to an encryption key, don't we have a function for that?

Reply:
Alright 👍🏻